File: examples/algorithms/src/leetcode/lc0226_invert_binary_tree.rs
/// LeetCode 226 - Invert Binary Tree
/// https://leetcode.com/problems/invert-binary-tree/
///
/// 题目：翻转一棵二叉树（镜像）。
///
/// 思路：递归后序遍历。先递归翻转左右子树，再交换当前节点的左右子节点指针。
/// 时间复杂度：O(n)
/// 空间复杂度：O(h)

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, Clone, PartialEq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn invert_tree(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
    if let Some(node) = root {
        let mut n = node.borrow_mut();
        let left = n.left.take();
        let right = n.right.take();
        n.left = invert_tree(right);
        n.right = invert_tree(left);
        drop(n);
        Some(node)
    } else {
        None
    }
}

// 辅助函数：中序遍历用于验证
fn inorder(root: &Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
    if let Some(n) = root {
        let n = n.borrow();
        inorder(&n.left, result);
        result.push(n.val);
        inorder(&n.right, result);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_invert_normal() {
        // 原树：
        //      4
        //     / \
        //    2   7
        //   / \ / \
        //  1  3 6  9
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 4,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 2,
                left: Some(Rc::new(RefCell::new(TreeNode::new(1)))),
                right: Some(Rc::new(RefCell::new(TreeNode::new(3)))),
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode {
                val: 7,
                left: Some(Rc::new(RefCell::new(TreeNode::new(6)))),
                right: Some(Rc::new(RefCell::new(TreeNode::new(9)))),
            }))),
        })));

        let inverted = invert_tree(root);
        let mut result = Vec::new();
        inorder(&inverted, &mut result);
        // 翻转后中序应为：9, 7, 6, 4, 3, 2, 1
        assert_eq!(result, vec![9, 7, 6, 4, 3, 2, 1]);
    }

    #[test]
    fn test_invert_empty() {
        assert_eq!(invert_tree(None), None);
    }

    #[test]
    fn test_invert_single() {
        let root = Some(Rc::new(RefCell::new(TreeNode::new(1))));
        let inverted = invert_tree(root);
        let mut result = Vec::new();
        inorder(&inverted, &mut result);
        assert_eq!(result, vec![1]);
    }
}

File: examples/by-algorithm/lc0226/python/lc0226_invert_binary_tree.py
"""
LeetCode 226. Invert Binary Tree
链接: https://leetcode.com/problems/invert-binary-tree/
难度: Easy

题目描述:
翻转一棵二叉树（镜像）。

形式化规约:
  输入: 二叉树根节点 root
  输出: 翻转后的二叉树根节点

最优解思路:
  递归后序遍历。先递归翻转左右子树，再交换当前节点的左右子节点。

复杂度:
  时间: O(n)
  空间: O(h)，h 为树高

正确性要点:
  1. 空树直接返回
  2. 先递归处理子树，确保子树已经翻转
  3. 交换左右指针完成当前节点的翻转
"""

from typing import Optional


class TreeNode:
    """二叉树节点定义。"""

    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    递归翻转二叉树。
    时间复杂度 O(n)，空间复杂度 O(h)。
    """
    if root is None:
        return None

    # 递归翻转左右子树
    left = invert_tree(root.left)
    right = invert_tree(root.right)

    # 交换左右子节点
    root.left = right
    root.right = left

    return root


def inorder_traversal(root: Optional[TreeNode]) -> list[int]:
    """中序遍历辅助函数，用于验证结果。"""
    result = []
    if root:
        result.extend(inorder_traversal(root.left))
        result.append(root.val)
        result.extend(inorder_traversal(root.right))
    return result


if __name__ == "__main__":
    # 示例 1:
    #      4
    #     / \
    #    2   7
    #   / \ / \
    #  1  3 6  9
    root = TreeNode(
        4,
        TreeNode(2, TreeNode(1), TreeNode(3)),
        TreeNode(7, TreeNode(6), TreeNode(9)),
    )
    inverted = invert_tree(root)
    assert inorder_traversal(inverted) == [9, 7, 6, 4, 3, 2, 1], "Example 1 failed"

    # 边界: 空树
    assert invert_tree(None) is None, "Empty tree failed"

    # 边界: 单节点
    single = TreeNode(1)
    assert inorder_traversal(invert_tree(single)) == [1], "Single node failed"

    print("All tests passed for LC 226 - Invert Binary Tree")
